Implemented add and subtract methods for CalcLength class

njlawton opened the pull request:
Implemented add and subtract as well as additional methods used to implement these methods: _convertToCalcLength and equals method

Commit:
added _convertToCalcLength method to LengthValue class
Method will return a CalcLength object equivilent of any child of class LengthValue that calls it i.e SimpleLength objects will return a CalcLength with one length type in it and CalcLength objects will return themselves

Commit:
Implemented _convertToCalcLength for CalcLength class
Implemented method such that a CalcLength will return itself added test case to verify the method

Commit:
Created simple unit test case for _convertToCalcLength() in SimpleLength class

Commit:
Implemented _convertToCalcLength for objects of class SimpleLength

Commit:
Implemented equals method in SimpleLength class
equals method checks wether the calling object and the input object are eqaul to each other this method will be used for testing and potentialy other uses in the future

Commit:
Implemented equals method in CalcLength class
equals method checks wether the calling object and the input object are eqaul to each other this method will be used for testing and potentialy other uses in the future

Commit:
Fixed faulty SimpleLength test case
Test case did not proporly check if two objects were equivilant and as such the test has been rewritten using a new equals method

Commit:
Implemented addition method for CalcLengths
Implementation has one simple test so far and needs more to ensure correct behaviour

Commit:
Added additional test cases for the equals method in CalcLength and SimpleLength

Commit:
Changed some variable names in simple test class to conform with code style
some variable names used short versions of words were as the code standard used requires variables use unshortened words

Commit:
Added test case for add method in CalcLength
Added test to ensure that CalcLenghts with different value types defind would still be correctly added in the new CalcLength

Commit:
Added unit tests for subtract method of CalcLength

Commit:
Implemented subtraction method for CalcLength class

Commit:
Fixed incorrect varible names in some CalcLength test cases

Commit:
Improved wording of some test cases relating to addition and subtraction
Changed some sentences to use consistant naming of classes across all tests, Fixed some spelling errors

Commit:
Fixed minor spelling mistakes CalcLength class

Commit:
Small improvements to commenting in the SimpleValue class

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+    
+    var calcDictionary = {};
+
+    //Iterate through all possible length types and add there values
```
Nitpicks: 
- Add single space after //
- s/there/their

Same below

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+    var calcDictionary = {};
+
+    //Iterate through all possible length types and add there values
+    for(var i = 0; i < shared.LengthValue.LengthType.length; i++){
```
Nitpick: need space after "for", and space between the closing ) and opening {

for (var i = 0; ...) {

Same elsewhere in this file

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+    //iterate through all length types and check that both objects contain the same values
+    for(var index in shared.LengthValue.LengthType){
+      var type = shared.LengthValue.LengthType[index];
+      if(this[type] != lengthValue[type]){
```
Also need space after if

if (condition ...

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+    for(var i = 0; i < shared.LengthValue.LengthType.length; i++){
+      var type = shared.LengthValue.LengthType[i];
+      if (this[type] == null && subtractedLength[type] == null) {
+        calcDictionary[type] = null;
```
In the constructor, you check if things are undefined, so it's fine to skip this case and end up with calcDictionary[type] being undefined. In this case, it's clearer to invert the conditions in the if statements:

if (subtractedLength[type] != null && this[type] != null) {
  calcDictionary[type] = this[type] - subtractedLength[type];
} else if (this[type] != null) {
  calcDictionary[type] = this[type];
} else if (subtractedLength[type] != null) {
  calcDictionary[type] = -subtractedLength[type];
}

Commit:
Minor improvements to code layout in the CalcLength class

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+    //Iterate through all possible length types and add there values
+    for(var i = 0; i < shared.LengthValue.LengthType.length; i++){
+      var type = shared.LengthValue.LengthType[i];
+      if (this[type] == null){
```
It's clearer to (find the things that are defined and add them) than it is to (find the things that are undefined and ad the other thing).

if (addedLength[type] != null && this[type] != null) {
calcDictionary[type] = this[type] + addedLength[type];
} else if (this[type] != null) {
calcDictionary[type] = this[type];
} else if (addedLength[type] != null) {
calcDictionary[type] = addedLength[type];
}

wilddamon commented on src/calc-length.js:
```diff
@@ -93,6 +93,72 @@
+      } else if (this[type] == null){
+        calcDictionary[type] = -subtractedLength[type];
+      } else if (subtractedLength[type] == null) {
+        calcDictionary[type] = -this[type];
```
This shouldn't be negative.

wilddamon commented on src/length-value.js:
```diff
@@ -50,13 +50,25 @@
+
+      return lengthToAdd_1._addCalcLengths(lengthToAdd_2);
+    }
     throw new TypeError('Not implemented yet');
```
This line can't be reached any more and can be deleted.

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
     return new SimpleLength((this.value - subtractedLength.value), this.type);
   };
 
+  SimpleLength.prototype._convertToCalcLength = function() {
```
I think it's more common to use _asCalcLength - plus it's shorter! :)

wilddamon commented on src/length-value.js:
```diff
@@ -50,13 +50,25 @@
+      var lengthToAdd_1 = this._convertToCalcLength();
+      var lengthToAdd_2 = addedLength._convertToCalcLength();
+
+      return lengthToAdd_1._addCalcLengths(lengthToAdd_2);
```
If you change _convertToCalcLength to _asCalcLength, it's fine to just write

return (this._asCalcLength())._addCalcLengths(addedLength._asCalcLength());

Similar below for subtract.

wilddamon commented on src/length-value.js:
```diff
@@ -73,6 +85,12 @@
     throw new TypeError('Not implemented yet');
   };
 
+  LengthValue.prototype._convertToCalcLength = function() {
```
This should probably throw an error saying it shouldn't be reached.

wilddamon commented on src/length-value.js:
```diff
@@ -73,6 +85,12 @@
+  LengthValue.prototype._convertToCalcLength = function() {
+  };
+
+  LengthValue.prototype.equals = function(lengthValue) {
```
Same here, should probably never be reached.

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
+    if (!(this.type == lengthValue.type && this.value == lengthValue.value)){
+      return false;
+    }
+      return true; 
```
Fix indenting

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
+  }
+
+  SimpleLength.prototype.equals = function(lengthValue) {
+    if (!(lengthValue instanceof SimpleLength)){
```
Needs space before {. Same below.

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
+    if (!(lengthValue instanceof SimpleLength)){
+      return false;
+    }
+    if (!(this.type == lengthValue.type && this.value == lengthValue.value)){
```
It's shorter and clearer to write

if (this.type != lengthValue.type || this.value != lengthValue.value)

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
+    return new CalcLength(calcDictionary);
+  }
+
+  SimpleLength.prototype.equals = function(lengthValue) {
```
It's more common to call the argument to .equals "other":

SimpleLength.prototype.equals = function(other) {

Same in the other files

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
   });
+
+  test('Adding two CalcLengths returns a new CalcLength with expected values in each value type', function() {
+    var cLength_1 = new CalcLength({px: 15, em: 6})
```
shouldn't contract names, so variables should be

var calcLength1
var calcLength2

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+    var cLength_1 = new CalcLength({px: 15, em: 6})
+    var cLength_2 = new CalcLength({px: 10, em: 3})
+    var lengthAddition = cLength_1.add(cLength_2);
+    var expectedlengthAddition = new CalcLength({px: 25, em: 9})
```
just call this "expected". Same elsewhere where you make a variable that stores an expected result. (You can also use "expectedResult" if you think that's clearer.

This is also a bit nitpicky, but it's often clearer to define all the necessary variables first, then construct the expectation, then do the function call under test, and finally do the assertions. So this function would look like:

{
var calcLength1 = new CalcLength({px: 15, em: 6});
var calcLength2 = new CalcLength({px: 10, em: 3});
var expectedResult = new CalcLength({px: 25, em: 9});

var result = calcLength1.add(calcLength2);

assert.isTrue(result.equals(expectedResult));
}

Also, you can reasonably assume that .equals returns false if the result and expected result have different types, so you can delete the line asserting that the result is a CalcLength.

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+  test('Adding two CalcLengths returns a new CalcLength with expected values in each value type', function() {
+    var cLength_1 = new CalcLength({px: 15, em: 6})
+    var cLength_2 = new CalcLength({px: 10, em: 3})
+    var lengthAddition = cLength_1.add(cLength_2);
```
just call this "result". Same elsewhere for variables that store the result for the test.

Commit:
Fixed implementation of subtraction method for CalcLength type
In old implemention a bug would occur that would make some values in the object calling the function negetive if those values were not present in the subtracting object this fixes this problem

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+  });
+
+    test('CalcLength do not have to have the same value types in order to be added and all value types seen in both'
+        +'CalcLengths with be in the outputted CalcLength', function() {
```
This is kind of long, how about

"All values in added CalcLengths appear in the output"

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+    assert.isTrue(expectedlengthSubtraction.equals(lengthSubtraction));
+  });
+
+    test('CalcLength do not have to have the same value types in order to be subtracted from each other and all value types '
```
Likewise, how about

"All values in subtracted CalcLengths appear in the output"

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+    assert.strictEqual(cLength, convertedCalcLength);
+  });
+
+  test('Test equals method of CalcLength such that it will return true if two CalcLength\'s are the same', function() {
```
Don't need the apostrophe.

How about:
'Equals method should return true for equal CalcLengths'

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+
+  test('Equals method should return true when one CalcLength has additional length types with null values', function() {
+  var cLength_1 = new CalcLength({px: 25, em: 3.2, percent: null})
+  var cLength_2 = new CalcLength({px: 25, em: 3.2})
```
These are represented the same internally so this test doesn't do anything and can be removed.

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+  assert.isTrue(cLength_1.equals(cLength_2));
+  });
+
+  test('Equals method should return false when one CalcLength has additional length types with non null values'
```
How about
"Equals method should return false for non-equal CalcLengths"?

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
     assert.strictEqual(lengthAddition.value, 5);
   });
 
+  test('convertToCalcLength method returns a CalcLength object with a single length value type when called from a SimpleLength object', function() {
```
We already know it's being called from a SimpleLength object since we're in the SimpleLength test suite. How about

"asCalcLength method returns a CalcLength with single value"?

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
+    assert.isTrue(convertedCalcLength.equals(testCalcLength));
+  });
+
+  test('Test equals method of Simple length such that it will return true if two simple lengths are the same', function() {
```
Similar to the calc length tests, how about

"Equals method should return true for equal SimpleLengths"

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
+  var simpleLength_2 = new CalcLength({em: 10});
+  
+  assert.isFalse(simpleLength_1.equals(simpleLength_2));
+  });
```
Please add a test for  "Equals method should return false for non-equal SimpleLengths" that checks .equals returns false if the value or the type is different.

Commit:
Improved add and subtract methods in the LengthValue class
methods will now throw a type error if a non LengthValue object is put into either method

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
+  assert.isTrue(simpleLength_1.equals(simpleLength_2));
+  });
+
+  test('Equals method should return false if a CalcLength is compared to a simple length even if they have the same value', function() {
```
How about "Equals method should return false if any other kind of object is passed in"

then you can also try

assert.isFalse(simpleLength.equals({}));
assert.isFalse(simpleLength.equals(5));

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
+
+  test('Equals method should return false if a CalcLength is compared to a simple length even if they have the same value', function() {
+  var simpleLength_1 = new SimpleLength(10, 'em');
+  var simpleLength_2 = new CalcLength({em: 10});
```
this isn't a SimpleLength.

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
+  test('convertToCalcLength method returns a CalcLength object with a single length value type when called from a SimpleLength object', function() {
+    var simpleLength = new SimpleLength(10, 'em');
+    var convertedCalcLength = simpleLength._convertToCalcLength();
+    var testCalcLength = new CalcLength({em: 10})
```
Call this "expected"

wilddamon commented on test/js/simple-length.js:
```diff
@@ -90,4 +90,24 @@ suite('SimpleLength', function() {
 
+  test('convertToCalcLength method returns a CalcLength object with a single length value type when called from a SimpleLength object', function() {
+    var simpleLength = new SimpleLength(10, 'em');
+    var convertedCalcLength = simpleLength._convertToCalcLength();
```
Call this "result"

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+    var lengthSubtraction = cLength_1.subtract(cLength_2);
+    var expectedlengthSubtraction = new CalcLength({px: 5, em: 3, percent: -5, ex: -6});
+
+    assert.instanceOf(lengthSubtraction, CalcLength, 'two added CalcLength of same type should be an instance of CalcLength');
```
Kind of verbose too. How something like "Result from adding two CalcLengths should be a CalcLength"?

wilddamon commented on test/js/calc-length.js:
```diff
@@ -83,4 +83,74 @@ suite('CalcLength', function() {
+  var cLength_2 = new CalcLength({px: 25, em: 3.2})
+  
+  assert.isTrue(cLength_1.equals(cLength_2));
+  });
```
It would also be good to check that .equals returns false if the passed in object is not a CalcLength.

wilddamon commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
+  SimpleLength.prototype._convertToCalcLength = function() {
+    var calcDictionary = {};
+
+    //Set calcDictionary field to the Length type inside the SimpleLength object that called the function
```
This comment is unnecessary. Please delete it.

Commit:
Renamed the _convertToCalcLength in LengthValue class to _asCalcLength

Commit:
Fixted minor indenting issue in the SimpleLength class

Commit:
Changed input variable name of equals method in LengthValue class
Variable name was changed to equals as this name is more common for equals methods

Commit:
Changed contracted variable names in CalcLenght and SimpleLength classes

Commit:
Shortened some test comments in the test suite to improve readability

Commit:
Minor changes to wording of test cases in SimpleLength test class

Commit:
Fixed a minor mistake in previous commit
Last commit changed the wrong comment in the test suite this commit fixes this problem

rjwright commented on src/calc-length.js:
```diff
@@ -93,6 +93,70 @@
+    }
+
+    // Iterate through all length types and check that both objects contain the same values
+    for (var index in shared.LengthValue.LengthType){
```
'index' should just be 'i'. Same elsewhere in this file. Loop counters can be given special names when there's a good reason to do so; otherwise just use i, j, k.

rjwright commented on src/calc-length.js:
```diff
@@ -69,9 +69,9 @@
+    for (var i = 0; i < shared.LengthValue.LengthType.length; i++){
       var type = shared.LengthValue.LengthType[i];
-      if(this[type] != null){
+      if (this[type] != null){
```
Missing space between bracket and brace here. Also on line 86, 88, 104, 106, 126, 130, 145, 150, 152.

rjwright commented on src/calc-length.js:
```diff
@@ -69,9 +69,9 @@
 
     // Iterate through all length types and multiply all non null lengths 
-    for(var i = 0; i < shared.LengthValue.LengthType.length; i++){
+    for (var i = 0; i < shared.LengthValue.LengthType.length; i++){
```
Missing space between bracket and brace here. Also on line 74, 86, 88, 104, 106, 126, 130, 145, 150, 152.

rjwright commented on src/calc-length.js:
```diff
@@ -93,6 +93,70 @@
 
+  CalcLength.prototype._addCalcLengths = function(addedLength) {
+    if (!(addedLength instanceof CalcLength)) {
+      throw new TypeError('Objects not of type CalcLength');
```
Maybe make the message 'Argument must be a CalcLength'. I assume this was copied from SimpleLength; feel free to also change it there.

rjwright commented on src/length-value.js:
```diff
@@ -50,15 +50,29 @@
+      var lengthToAdd_1 = this._asCalcLength();
+      var lengthToAdd_2 = addedLength._asCalcLength();
+
+      return lengthToAdd_1._addCalcLengths(lengthToAdd_2);
```
I think Eddy might have mentioned this in a previous review. You can just say

```
return this._asCalcLength()._addCalcLengths(addedLength._asCalcLength());
```

Also, for future reference, please don't use underscores in any identifier names except for object members that you wish to be treated as private. E.g. here you would use lengthToAdd1, lengthToAdd2.

rjwright commented on src/length-value.js:
```diff
@@ -50,15 +50,29 @@
+      var lengthToAdd_1 = this._asCalcLength();
+      var lengthToAdd_2 = subtractedLength._asCalcLength();
+
+      return lengthToAdd_1._subtractCalcLengths(lengthToAdd_2);
```
Again, no need to make lengthToAdd\* variables.

rjwright commented on src/length-value.js:
```diff
@@ -50,15 +50,29 @@
+
+      return lengthToAdd_1._addCalcLengths(lengthToAdd_2);
+    } else {
+      throw new TypeError('The object you are adding is not of type LengthValue');
```
Make this error message consistent with the ones in calc-length.js

rjwright commented on src/length-value.js:
```diff
@@ -50,15 +50,29 @@
+
+      return lengthToAdd_1._subtractCalcLengths(lengthToAdd_2);
+    } else {
+      throw new TypeError('The object you are trying to subtract from this length is not of type LengthValue');
```
Make this error message consistent with the ones in calc-length.js

rjwright commented on src/simple-length.js:
```diff
@@ -67,6 +67,25 @@
 
+  SimpleLength.prototype._asCalcLength = function() {
+    var calcDictionary = {};
+
```
These blank lines are unnecessary. I think it's good to use whitespace to break up blocks (exactly when is a matter of taste), but for very small blocks you should just bunch them up.

Commit:
Fixed formatting issues found using gjslint for the ClacLength testing class

Commit:
Fixed formatting issues found using gjslint for the SimpleLength testing class

Commit:
Fixed formatting issues found using gjslint for the CalcLength and LengthValue class

Commit:
Fixed formatting issues found using gjslint for the SimpleLength class

Commit:
Changed type error statements in Simple and CalcLength class to improve understandability

Commit:
Changed some expretions to reduse the number of unnesassary variables created

Commit:
Removed some whitespace to improve readability in the SimpleLength class

wilddamon commented on src/calc-length.js:
```diff
@@ -90,6 +90,70 @@
+    for (var i = 0; i < shared.LengthValue.LengthType.length; i++) {
+      var type = shared.LengthValue.LengthType[i];
+      if (this[type] == null) {
+        calcDictionary[type] = addedLength[type];
```
You need to check if both are defined before you do this test case. If you do it this way around, addedLength[type] can still be null here. This sort of works, because
1 + null = null
null + 2 = 2

however
null + null = 0

So you will potentially end up with extra zeros in your output calclength.

This is why it's easier to check if things are defined rather than whether they are undefined.

Same below.

wilddamon commented on src/length-value.js:
```diff
@@ -50,15 +50,23 @@
     if (this instanceof SimpleLength && addedLength instanceof SimpleLength && this.type == addedLength.type) {
       return this._addSimpleLengths(addedLength);
+    } else if (addedLength instanceof LengthValue) {
+      //ensure both lengths are of type CalcLength before adding
```
Nitpick: Comment should start with capital letter, need space after //. Same for subtract.

wilddamon commented on src/length-value.js:
```diff
@@ -73,6 +81,12 @@
     throw new TypeError('Not implemented yet');
   };
 
+  LengthValue.prototype._asCalcLength = function() {
```
throw new TypeError("Should not be reached"). Same for equals

wilddamon commented on src/simple-length.js:
```diff
@@ -64,6 +64,23 @@
 
+  SimpleLength.prototype._asCalcLength = function() {
+    var calcDictionary = {};
+    //Set calcDictionary field to the Length type inside the SimpleLength object that called the function
```
Nitpick: Need space after //.

wilddamon commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,94 @@ suite('CalcLength', function() {
+  test('Adding two CalcLengths returns a new CalcLength with expected values in each value type', function() {
+    var calcLength1 = new CalcLength({px: 15, em: 6});
+    var calcLength2 = new CalcLength({px: 10, em: 3});
+    var lengthAddition = calcLength1.add(calcLength2);
```
Just call this "result". Same for similar variables elsewhere. It makes it clearer that this is the result of an operation.

wilddamon commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,94 @@ suite('CalcLength', function() {
+    var calcLength1 = new CalcLength({px: 15, em: 6, percent: 5});
+    var calcLength2 = new CalcLength({px: 10, em: 3, ex: 6});
+    var lengthAddition = calcLength1.add(calcLength2);
+    var expectedlengthAddition = new CalcLength({px: 25, em: 9, percent: 5, ex: 6});
```
result and expectedResult. It's more common so it's easier to read at a glance.

Commit:
Improvements to the add and subtract methods for the CalcLength class
Methods now check if for a certain length value both objects have it set to null this will avoid many subtile errors due to adding nulls together

Commit:
Minor changes in the LengthValue class
Improved some commenting and removed some obsolete errors

Commit:
Changed some variable names in test cases to improve readability
Changed many variable names to result and expected result as thease are more commain and easier to read

rjwright commented on src/calc-length.js:
```diff
@@ -90,6 +90,74 @@
+    }
+
+    // Iterate through all length types and check that both objects contain the same values
+    for (var index in shared.LengthValue.LengthType) {
```
'index' should just be 'i'. Same elsewhere in this file. Loop counters can be given special names when there's a good reason to do so; otherwise just use i, j, k.

rjwright commented on src/length-value.js:
```diff
@@ -50,27 +50,39 @@
+      // Ensure both lengths are of type CalcLength before adding
+      return this._asCalcLength()._addCalcLengths(addedLength._asCalcLength());
+    } else {
+      throw new TypeError('The object you are adding is not of type LengthValue');
```
Make this error message consistent with the ones in calc-length.js

rjwright commented on src/length-value.js:
```diff
@@ -50,27 +50,39 @@
+      // Ensure both lengths are of type CalcLength before adding
+      return this._asCalcLength()._subtractCalcLengths(subtractedLength._asCalcLength());
+    } else {
+      throw new TypeError('The object you are trying to subtract from this length is not of type LengthValue');
```
Make this error message consistent with the ones in calc-length.js

rjwright commented on test/js/simple-length.js:
```diff
@@ -46,48 +46,66 @@ suite('SimpleLength', function() {
+  });
+
+  test('Equals method should return true for equal SimpleLengths', function() {
+  var simpleLength1 = new SimpleLength(10, 'em');
```
Indenting. Same with next testcase.

rjwright commented on src/simple-length.js:
```diff
@@ -64,6 +64,23 @@
 
+  SimpleLength.prototype._asCalcLength = function() {
+    var calcDictionary = {};
+    // Set calcDictionary field to the Length type inside the SimpleLength object that called the function
```
Unnecessary comment.

rjwright commented on src/length-value.js:
```diff
@@ -50,27 +50,39 @@
+    throw new TypeError('Should not be reached');
+  };
+
+  LengthValue.prototype._asCalcLength = function() {
```
As previously requested by Eddy: "throw new TypeError("Should not be reached"). Same for equals"

Commit:
Fixed small issues in Lenght classes
Changed some statements e.g. for loop types to be more appropriate for the useadded some additinal errors to clarify the use of some methods in the LengthValue class

rjwright commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,114 @@ suite('CalcLength', function() {
+    var result = calcLength1.add(calcLength2);
+    var expectedResult = new CalcLength({px: 25, em: 9, percent: 5, ex: 6});
+
+    assert.instanceOf(result, CalcLength, 'two added CalcLength of same type should be an instance of CalcLength');
```
Not sure what you mean by "of the same type". Clearly they're both CalcLengths, but they hold differently shaped data.

Also s/CalcLength/CalcLengths. Same throughout file.

rjwright replied:
And actually we don't really need all of these instanceOf asserts for add etc. They are repetitive, but also LengthValue.equals checks types. You can leave in one for each method (one for add, one for subtract, ect.) for the sake of explicitness if you want. The rest can go.

rjwright replied:
And if you cut it down just one per method, change the error message to "[operation]ing two CalcLengths returns a CalcLength".

rjwright commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,114 @@ suite('CalcLength', function() {
+    assert.isTrue(expectedResult.equals(result));
+  });
+
+  test('length values in CalcLengths set to null add to null not zero', function() {
```
Maybe "Null length values in CalcLengths add to null, not zero".

rjwright commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,114 @@ suite('CalcLength', function() {
+    assert.isTrue(expectedResult.equals(result));
+  });
+
+  test('length values in CalcLengths set to null when subtracted will equal null not zero', function() {
```
"Subtracting null length values in CalcLengths results in null, not zero"?

rjwright commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,114 @@ suite('CalcLength', function() {
+    assert.strictEqual(calcLength, result);
+  });
+
+  test('Test equals method of CalcLength such that it will return true if two CalcLength\'s are the same', function() {
```
"CalcLength.equals returns true if the compared CalcLengths are the same."

rjwright commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,114 @@ suite('CalcLength', function() {
+  });
+
+  test('Test equals method of CalcLength such that it will return true if two CalcLength\'s are the same', function() {
+  var calcLength1 = new CalcLength({px: 25, em: 3.2});
```
Indenting. And below.

rjwright commented on test/js/simple-length.js:
```diff
@@ -46,48 +46,66 @@ suite('SimpleLength', function() {
+    var simpleLength1 = new SimpleLength(10, 'em');
+    var simpleLength2 = new SimpleLength(5, 'em');
+    var result = simpleLength1.add(simpleLength2);
+    assert.instanceOf(result, SimpleLength, 'two added SimpleLength of same type should be an instance of SimpleLength');
```
s/SimpleLength/SimpleLengths. And below.

rjwright commented:
LGTM when comments on test files are resolved. Will wait for LGTM from @wilddamon  before landing.

Commit:
Fixed issues raised in reveiw in the CalcLength test file

Commit:
Fixed issues raised in reveiw in the SimpleLength test file

rjwright commented on src/calc-length.js:
```diff
@@ -90,6 +90,74 @@
+    return new CalcLength(calcDictionary);
+  };
+
+  LengthValue.prototype._asCalcLength = function() {
```
This needs to be CalcLength.prototype._asCalcLength, right?

Commit:
Fixed error in CalcLength class and added additinal test cases
Fixed an error not discovered during testing in the CalcLength class also added some test cases to the addition and subtraction methods for the SimpleLength class to ensure that adding SimpleLengths of different types would return a CalcLength

Commit:
Changed words used in some test cases as some method names have changed

wilddamon commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,113 @@ suite('CalcLength', function() {
+    var result = calcLength1.add(calcLength2);
+    var expectedResult = new CalcLength({px: 25, em: null, percent: 5, ex: 6});
+
+    assert.isTrue(expectedResult.equals(result));
```
In this case it is better to check that 

assert.strictEqual(null, result.em)

wilddamon commented on test/js/calc-length.js:
```diff
@@ -61,26 +61,113 @@ suite('CalcLength', function() {
+    var result = calcLength1.subtract(calcLength2);
+    var expectedResult = new CalcLength({px: 5, em: null, percent: 5, ex: -6});
+
+    assert.isTrue(expectedResult.equals(result));
```
Likewise here...

wilddamon commented:
Two very minor comments, otherwise LGTM

Commit:
Changed the asserted statement to check for null values in CalcLength test class
